fastruby/pecas
# Feature/entry validation alerts

Closes #98

**Description:**
Once a day, we should have a scheduled job that reviews all the entries submitted by the team during that day. The task should probably run every weekday at 8pm in the user's local time zone as defined by Slack.

Initially only internal employees would receive these messages. We determine which users to include by inspecting the `@ombuteam` on Slack. 

The job would check that the entries are well formed. They should meet this criteria: 

- They should have a description that is more than two words. 
- If the entry doesn't have a #calls label, then it should have a JIRA ID. 
- If the entry doesn't have a JIRA ID or URL, then it should notify the person in Slack

The notification should include a list of entries that look suspicious. Here is a template: 

```
@etagwerker Hi Ernesto, I found these entries that are missing a JIRA ticket or ID: 

- Did work (3 hours) 
- Worked on stuff (1 hour) 

Please make sure that these entries are accurate. If they are all good, great! If not, please make sure you fix them so you don't get penalized for any of them!
```

**To Do**
* [x] Create and permission a Slack Bot
* [x] Build a Slack Service to utilize the Web API of the Slack Bot
* [x] Build a rules system to assess time entry descriptions
* [x] Build a TimeEntry domain to manage finding the correct Slack Users, current entries, and validation logic
* [x] Build a rake task to utilize this domain
* [x] Set up a cron job to call the rake task on a 1 hour interval

QA
* tests pass
* Add yourself to the pecas_dev group on Slack
* Add a time entry for today that lacks a Jira ticket, url, or the #calls tag
* spin up the app between 7pm and 8pm
* You should get a message on Slack an hour late.

<img width="663" alt="Screen Shot 2022-03-02 at 5 59 11 PM" src="https://user-images.githubusercontent.com/607860/156485083-3de5b5ee-eec9-4052-a89e-ee3a175579c9.png">



I will abide by the [code of conduct](code_of_conduct.md).
